File: standard/src/main/java/com/KillerBLS/modpeide/adapter/SuggestionAdapter.java
/*
 * Licensed to the Light Team Software (Light Team) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The Light Team licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.KillerBLS.modpeide.adapter;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.TextView;

import com.KillerBLS.modpeide.R;
import com.KillerBLS.modpeide.adapter.model.SuggestionItem;

import java.util.ArrayList;
import java.util.Collection;

/**
 * @author Trần Lê Duy
 */
public class SuggestionAdapter extends ArrayAdapter<SuggestionItem> {

    private ArrayList<SuggestionItem> mData; //Полный список
    private ArrayList<SuggestionItem> mDataFiltered; //Отображаемый список

    private Filter mFilter = new Filter() {
        @Override
        public CharSequence convertResultToString(Object resultValue) {
            if (resultValue == null) {
                return "";
            }
            return ((SuggestionItem) resultValue).getName();
        }

        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            FilterResults filterResults = new FilterResults();
            mDataFiltered.clear();
            if (constraint != null) {
                for (SuggestionItem item : mData) {
                    if (item.compareTo(constraint.toString()) == 0) {
                        mDataFiltered.add(item);
                    }
                }
                filterResults.values = mDataFiltered;
                filterResults.count = mDataFiltered.size();
            }
            return filterResults;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            ArrayList<SuggestionItem> filteredList = (ArrayList<SuggestionItem>) results.values;
            clear();
            if (filteredList != null && filteredList.size() > 0) {
                addAll(filteredList);
            }
            notifyDataSetChanged();
        }
    };

    // region BASE

    public SuggestionAdapter(@NonNull Context context,
                             @LayoutRes int resource, @NonNull ArrayList<SuggestionItem> objects) {
        super(context, resource, objects);
        mData = (ArrayList<SuggestionItem>) objects.clone();
        mDataFiltered = new ArrayList<>();
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        if (convertView == null) {
            convertView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_suggestion, null);
        }
        final SuggestionItem item = getItem(position);
        if(item != null) {
            TextView title = convertView.findViewById(R.id.item_title);
            title.setText(item.getName());
        }
        return convertView;
    }

    @NonNull
    @Override
    public Filter getFilter() {
        return mFilter;
    }

    // endregion BASE

    // region METHODS

    public ArrayList<SuggestionItem> getAllItems() {
        return mData;
    }

    public void clearAllData() {
        super.clear();
        mData.clear();
    }

    public void addData(@NonNull Collection<? extends SuggestionItem> collection) {
        addAll(collection);
        mData.addAll(collection);
    }

    // endregion METHODS
}

File: editorkit/src/main/kotlin/com/lightteam/editorkit/adapter/SuggestionAdapter.kt
/*
 * Licensed to the Light Team Software (Light Team) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The Light Team licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lightteam.editorkit.adapter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import com.lightteam.editorkit.feature.colorscheme.ColorScheme
import com.lightteam.editorkit.feature.suggestions.WordsManager
import com.lightteam.language.model.SuggestionModel

abstract class SuggestionAdapter(
    context: Context,
    resourceId: Int
) : ArrayAdapter<SuggestionModel>(context, resourceId) {

    var colorScheme: ColorScheme? = null
    var wordsManager: WordsManager? = null

    private var queryText = ""

    abstract fun createViewHolder(parent: ViewGroup): SuggestionViewHolder

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val viewHolder = createViewHolder(parent)
        viewHolder.bind(getItem(position), queryText)
        return viewHolder.itemView
    }

    override fun getFilter(): Filter {
        return object : Filter() {

            private val suggestions: MutableList<SuggestionModel> = mutableListOf()

            override fun performFiltering(constraint: CharSequence?): FilterResults {
                val filterResults = FilterResults()
                suggestions.clear()
                wordsManager?.let {
                    val query = constraint.toString()
                    for (suggestion in it.suggestions) {
                        val word = suggestion.text.toString()
                        if (word.startsWith(query, ignoreCase = true) &&
                            !word.equals(query, ignoreCase = true)) {
                            queryText = query
                            suggestions.add(suggestion)
                        }
                    }
                }
                filterResults.values = suggestions
                filterResults.count = suggestions.size
                return filterResults
            }

            override fun publishResults(constraint: CharSequence?, results: FilterResults) {
                clear()
                addAll(suggestions)
                notifyDataSetChanged()
            }
        }
    }

    abstract class SuggestionViewHolder(val itemView: View) {
        abstract fun bind(suggestion: SuggestionModel?, query: String)
    }
}